gpcontrib/arenadata_toolkit/arenadata_toolkit.c

```diff
@@ -1,4 +1,198 @@
+#include <dirent.h>
+#include <sys/stat.h>
+#include <ctype.h>
+
 #include "postgres.h"
 #include "fmgr.h"
+#include "funcapi.h"
+#include "miscadmin.h"
+#include "storage/fd.h"
+#include "nodes/execnodes.h"
+#include "cdb/cdbvars.h"
+#include "utils/builtins.h"
+#include "utils/relfilenodemap.h"
+#include "utils/timestamp.h"
+#include "catalog/pg_tablespace.h"
 
 PG_MODULE_MAGIC;
+
+typedef struct
+{
+	char     *datpath;
+	DIR      *dirdesc;
+	TupleDesc tupdesc;
+} user_fctx_data;
+
+static bool check_equal_substring(const char *str, const char *sub,
+								  int startPos, int checkLen)
+{
+	for (int pos = 0; pos < checkLen; pos++)
+	{
+		if (tolower(str[startPos + pos]) == sub[pos])
+			continue;
+
+		return false;
+	}
+
+	return true;
+}
+
+static bool should_skip_file(const char *filename)
+{
+	/*
+	 * Must use strings with lower case symbols.
+	 * NULL must be at the end of arrays.
+	 */
+	const char *prefixes[] =
+	{
+		"pg", "t_", NULL
+	};
+	const char *postfixes[] =
+	{
+		"_vm", "_fsm", "_init", NULL
+	};
+
+	int filenamelen = strlen(filename);
+	for (int idx = 0; prefixes[idx] != NULL; idx++)
+	{
+		const char * prefix = prefixes[idx];
+		int plen = strlen(prefix);
+		if (filenamelen - plen < 0)
+			continue;
+
+		if (check_equal_substring(filename, prefix, 0, plen))
+			return true;
+	}
+
+	for (int idx = 0; postfixes[idx] != NULL; idx++)
+	{
+		const char * postfix = postfixes[idx];
+		int plen = strlen(postfix);
+		if (filenamelen - plen < 0)
+			continue;
+
+		int startPos = filenamelen - plen;
+		if (check_equal_substring(filename, postfix, startPos, plen))
```

Comment:
Congratulations! You invented the wheel instead of `pg_strncasecmp`!

Reply:
It's better to use the original version https://github.com/arenadata/gpdb/blob/db2d3880fb989ef49afa121d13869f768fd2c901/gpcontrib/gp_replica_check/gp_replica_check.c#L44-L49

Reply:
It was the first variant in my patch, you said about it:
```
Why is this macro needed?
Why not just put these conditions in functions?
```
Could you choose variant which I can use? Or I will be fixing it for a long time.

Reply:
To improve performance, I proposed an option where there were fewer function calls.